Update documentation with new features and instructions for new blocks

menon-karthik opened the pull request:
<!-- Give your pull request (PR) a descriptive name -->

## Current situation
Closes #66 
We recently had a few new features implemented (or in soon-to-be-merged PRs). So I made some updates to the documentation to reflect those changes. 


## Release Notes 
* Added a new page documenting how to add new blocks.
* Added a new section on contributing to the code.
* Added documentation on #108 
* Added documentation on #109 
* Added short summary of available blocks (including new ones)


## Documentation
This is the documentation.


## Testing
Documentation builds. Did not change anything else.


## Code of Conduct & Contributing Guidelines 
- I agree to follow the [Code of Conduct](https://github.com/SimVascular/.github/blob/main/CODE_OF_CONDUCT.md) and [Contributing Guidelines](https://github.com/SimVascular/.github/blob/main/CONTRIBUTING.md).

menon-karthik committed:
added cycle to cycle error and exprtk functionailty to main.md. initial version of docs to add new blocks

menon-karthik committed:
typo in microvascular

menon-karthik committed:
cleaning up new block info

menon-karthik committed:
test formatting for new blocks

menon-karthik committed:
added equation example

menon-karthik committed:
new file for add_block

menon-karthik committed:
test spacing and clean up developer guide

menon-karthik committed:
test formatting

menon-karthik committed:
bug

menon-karthik committed:
test_spacing

menon-karthik committed:
added spacing

menon-karthik committed:
minor edits

menon-karthik committed:
test_space

menon-karthik committed:
cleaned up

menon-karthik committed:
subheadings

menon-karthik committed:
modified subheadings

menon-karthik committed:
subheading foix

menon-karthik committed:
cleaned up subheadings

menon-karthik committed:
added contributing section

menon-karthik committed:
test link section

menon-karthik committed:
clean up

mrp089 commented on docs/pages/main.md:
```diff
@@ -314,13 +326,17 @@ Blood vessel \n junction              | MODEL::BloodVesselJunction  | `BloodVess
 ------------------------------------- | --------------------------- | --------------------- | ----------- 
-Prescribed (transient) flow           | MODEL::FlowReferenceBC      | `FLOW`                | `Q`: Time-dependent flow values \n `t`: Time stamps
-Prescribed (transient) pressure       | MODEL::PressureReferenceBC  | `PRESSURE`            | `P`: Time-dependent pressure values \n `t`: Time stamps
+Prescribed (transient) flow           | MODEL::FlowReferenceBC      | `FLOW`                | `Q`: Time-dependent flow values \n `t`: Time steps \n `fn`: Mathematical expression
```
Indicate that it's either (`Q` and `t`) or `fn`

mrp089 commented on docs/pages/main.md:
```diff
@@ -314,13 +326,17 @@ Blood vessel \n junction              | MODEL::BloodVesselJunction  | `BloodVess
+
+The above table describes the most commonly used boundary conditions. In addition, svZeroDSolver includes various closed-loop boundary conditions. Examples can be found in `svZeroDSolver/tests/cases`.
 
+Note that the `FLOW` and `PRESSURE` boundary conditions accept mathematical expressions in `bc_values`. For an example where values of the boundary condition are specified as a function of time, see `svZeroDSolver/tests/cases/pulsatileFlow_R_RCR.json`. For an example with a mathematical expression for the boundary condition, see `svZeroDSolver/tests/cases/timeDep_Flow.json`. 
```
Might be easier to read if the examples were included here (and easier to see for new users that mathematical expressions are possible).

mrp089 commented on docs/pages/developer_guide.md:
```diff
@@ -30,15 +26,35 @@ cmake -DCMAKE_BUILD_TYPE=Debug ..
+
+* The changes should be implemented in a feature branch of the developer's fork of svZeroDSolver. 
+* Once the changes are implemented, the developer should make sure the build, documentation, and code format tests are passing on the user's feature branch. 
+  * The tests are automatically run when pushing changes to the developer's remote branch on Github. 
```
Good thing to write this down

mrp089 commented on docs/pages/add_block.md:
````diff
@@ -0,0 +1,159 @@
+
+* The new class will be inherited from `Block`. Define a constructor of the form:
+```
+    GenericBlock(int id, Model *model) 
````
Maybe something like `MyNewBlock` instead of `GenericBlock`

mrp089 commented on docs/pages/add_block.md:
```diff
@@ -0,0 +1,159 @@
+  * `block_type` and `block_class` are the same as what was added in Step 1 above.
+  * The names of the input parameters of the block are `Param_1`, ... , `Param_N`. 
+  * The properties of each parameter are defined by `InputParameter`, which specifies whether it is optional, an array, a scalar, a function, and its default value.
+  * The names `Param_1`, ... , `Param_N` should be the same as the parameter names within the block definition in the `.json` configuration/input file.
```
should or have to?

mrp089 commented on docs/pages/add_block.md:
```diff
@@ -0,0 +1,159 @@
+<p> <br> </p>
+
+### Set up the degrees of freedom
+* The class should have a `setup_dofs(DOFHandler &dofhandler)` function.
```
should or has to?

mrp089 commented on docs/pages/add_block.md:
````diff
@@ -0,0 +1,159 @@
+a*dQ_in/dt + b*P_in + c*(dP_in/dt)*Q_in + d = 0
+```
+```
+e*dP_out/dt + f*Q_out*Q_out + g*P_out + h*I_1 = 0
````
Write in math mode?

mrp089 commented on docs/pages/add_block.md:
```diff
@@ -0,0 +1,159 @@
+  * All matrix elements that are constant are specified in `update_constant`.
+  * Matrix elements that depend only on time (not the state variables) are specified in `update_time`.
+  * Matrix elements that change with the solution (i.e. depend on the state variables themselves) are specified in `update_solution`. 
+  * *Note: Not all blocks will require the `update_time` and `update_solution` functions.*
```
Would change the language above from "should" to something like "must"

mrp089 commented on docs/pages/add_block.md:
```diff
@@ -0,0 +1,159 @@
+
+## 4. Add the new block to the build system. 
+
+* Add `GenericBlock.h` and `GenericBlock.cpp` to `src/model/CMakeLists.txt`
```
Would consistently move something like `MyNewBlock`

mrp089 reviewed (changes requested):
Looks very nice! Thank you for updating the documentation! I added a few minor ideas.

menon-karthik committed:
review comments

menon-karthik committed:
test eqn

menon-karthik committed:
test eqn again

menon-karthik committed:
mod eqn

mrp089 committed:
GitHub actions: remove old os, add latest

mrp089 commented:
The `macos-11` GitHub runner is [not available anymore](https://github.com/actions/runner-images/issues/9255) as of July, so I removed it.

I added a `macos-latest` runner (and `ubuntu-latest`) so we always stay up to date. However, there seems to be a problem with `conda`. Up to you, @menon-karthik, if you want to do this upgrade now or leave it for a later PR.

menon-karthik committed:
doxy latex font size

menon-karthik committed:
test conda PATH

menon-karthik committed:
try codna alias

menon-karthik committed:
added eqns, modified size

menon-karthik committed:
test conda-incubator in GH actions to find conda

menon-karthik committed:
bug in eqns

menon-karthik committed:
eqns clean

menon-karthik committed:
added outputs description, debug eqn

menon-karthik committed:
test square bracket

menon-karthik committed:
remove sq bracket

menon-karthik committed:
remove eq

menon-karthik committed:
add back eq

menon-karthik committed:
remove eq1

menon-karthik committed:
state vector

menon-karthik committed:
fix eqn 1

menon-karthik commented:
> The `macos-11` GitHub runner is [not available anymore](https://github.com/actions/runner-images/issues/9255) as of July, so I removed it.
> 
> I added a `macos-latest` runner (and `ubuntu-latest`) so we always stay up to date. However, there seems to be a problem with `conda`. Up to you, @menon-karthik, if you want to do this upgrade now or leave it for a later PR.

I think it's a good idea to stay ahead of the curve, so let's upgrade that right away. I fixed the `conda` problem by using [conda-incubator](https://github.com/conda-incubator/setup-miniconda) to find `conda` and add it to the path. Also, it looks like `macos-12` is also going to be removed soon. I will create a new issue and PR to upgrade that just so it is recorded.

mrp089 committed:
Merge branch 'master' into update_docs
